api/certs/cert.go

```diff
@@ -0,0 +1,218 @@
+	wssUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpWssCert, config.Parameters.HttpWssKey, config.Parameters.HttpWssDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	httpsUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpsJsonCert, config.Parameters.HttpsJsonKey, config.Parameters.HttpsJsonDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	defaultDomain, err := node.GetDefaultDomainFromIP(config.Parameters.Hostname, config.Parameters.DefaultTlsDomainTmpl)
+	if err != nil {
+		log.Error(err)
+	}
+	_, _, certFile, keyFile := GetACMEFileNames(defaultDomain, config.Parameters.CertDirectory)
+	defaultCertExists, err := VerifyLocalCertificate(certFile, keyFile, defaultDomain)
+	if err != nil {
+		log.Error(err)
+	}
+
+	return wssUserCertExists, httpsUserCertExists, defaultCertExists
+}
+
+func CheckCerts() (chan struct{}, chan struct{}) {
+	wssCertExists, httpsCertExists, defaultCertExists := CertExists()
+	ChangeToDefault(httpsCertExists, wssCertExists)
+
+	wssCertReady := make(chan struct{}, 1)
+	httpsCertReady := make(chan struct{}, 1)
+
+	if wssCertExists || defaultCertExists {
+		go func() {
+			wssCertReady <- struct{}{}
+		}()
+	}
+	if httpsCertExists || defaultCertExists {
+		go func() {
+			httpsCertReady <- struct{}{}
+		}()
+	}
+	if !(wssCertExists && httpsCertExists) && !defaultCertExists {
+		go PrepareCert(wssCertExists, httpsCertExists, wssCertReady, httpsCertReady)
+	}
+	return wssCertReady, httpsCertReady
+}
+
+func PrepareCert(wssCertExists, httpsCertExists bool, wssCertReady, httpsCertReady chan struct{}) {
+	first := true
+	for {
+		cm, err := NewCertManager()
+		if err != nil {
+			log.Errorf("create new certManager err: %v", err)
+			return
+		}
+		if !first {
+			time.Sleep(config.Parameters.CertCheckInterval * time.Second)
+		}
+		first = false
+		if cm.NeedObtain() {
+			cert, err := cm.ObtainNewCert()
+			if err != nil {
+				log.Errorf("apply cert failed: %v", err)
+				continue
+			}
+			err = saveCertToDisk(cert)
+			if err != nil {
+				log.Errorf("save cert failed: %v", err)
+				continue
+			}
+			if !wssCertExists {
+				select {
+				case wssCertReady <- struct{}{}:
+					log.Info("wss cert ready")
+				}
+			}
+			if !httpsCertExists {
+				select {
+				case httpsCertReady <- struct{}{}:
+					log.Info("https cert ready")
+				}
+			}
+		} else if cm.NeedRenew() {
+			renew, err := cm.RenewCert()
+			if err != nil {
+				log.Errorf("renew cert failed: %v", err)
+				continue
+			}
+			err = saveCertToDisk(renew)
+			if err != nil {
+				log.Errorf("save renew cert failed: %v", err)
+				continue
+			}
+		}
+	}
+}
+
+func ChangeToDefault(https, wss bool) {
+	defaultDomain, err := node.GetDefaultDomainFromIP(config.Parameters.Hostname, config.Parameters.DefaultTlsDomainTmpl)
+	if err != nil {
+		log.Error(err)
+	}
+
+	userFile, resourceFile, certFile, keyFile := GetACMEFileNames(defaultDomain, config.Parameters.CertDirectory)
+	config.Parameters.DefaultTlsCert = certFile
```

Comment:
Suggest to add an info log when config is changed:
Set default cert domain to <defaultDomain>